pkg/3scale/amp/operator/deployment_config_reconciler.go

```diff
@@ -4,228 +4,108 @@ import (
+		update = true
 	}
 
-	r.ensurePodContainers(updated.Spec.Template.Spec.InitContainers, desired.Spec.Template.Spec.InitContainers)
-	if !reflect.DeepEqual(updated.Spec.Template.Spec.InitContainers, desired.Spec.Template.Spec.InitContainers) {
-		fmt.Println(cmp.Diff(updated.Spec.Template.Spec.InitContainers, desired.Spec.Template.Spec.InitContainers, cmpopts.IgnoreUnexported(resource.Quantity{})))
-		updated.Spec.Template.Spec.InitContainers = desired.Spec.Template.Spec.InitContainers
-		changed = true
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.Containers[0].Resources, desired.Spec.Template.Spec.Containers[0].Resources) {
+		diff := cmp.Diff(existing.Spec.Template.Spec.Containers[0].Resources, desired.Spec.Template.Spec.Containers[0].Resources, cmpopts.IgnoreUnexported(resource.Quantity{}))
+		logger.Info(fmt.Sprintf("%s spec.template.spec.containers[0].resources have changed: %s", desiredName, diff))
+		existing.Spec.Template.Spec.Containers[0].Resources = desired.Spec.Template.Spec.Containers[0].Resources
+		update = true
 	}
 
-	r.ensureDeploymentConfigStrategy(&updated.Spec.Strategy, &desired.Spec.Strategy)
-	if !reflect.DeepEqual(updated.Spec.Strategy, desired.Spec.Strategy) {
-		fmt.Println(cmp.Diff(updated.Spec.Strategy, desired.Spec.Strategy, cmpopts.IgnoreUnexported(resource.Quantity{})))
-		updated.Spec.Strategy = desired.Spec.Strategy
-		changed = true
-	}
-
-	return changed, nil
+	return update
 }
 
-func (r *DeploymentConfigReconciler) ensureDeploymentConfigStrategy(updated, desired *appsv1.DeploymentStrategy) {
-	// When no DeploymentConfig Strategy Type is defined OpenShift defines a default one
-	// We make sure we have that into desired in that case so the diff does not
-	// detect that in the comparison
-	if desired.Type == appsv1.DeploymentStrategyType("") {
-		*desired = *updated.DeepCopy()
-	}
-
-	if desired.Type == appsv1.DeploymentStrategyTypeRolling && desired.RollingParams == nil {
-		desired.RollingParams = updated.RollingParams
-	}
-
-	if desired.Type == appsv1.DeploymentStrategyTypeRecreate && desired.RecreateParams == nil {
-		desired.RecreateParams = updated.RecreateParams
-	}
-
-	if desired.Type == appsv1.DeploymentStrategyTypeCustom && desired.CustomParams == nil {
-		desired.CustomParams = updated.CustomParams
-	}
-
-	if desired.ActiveDeadlineSeconds == nil {
-		desired.ActiveDeadlineSeconds = updated.ActiveDeadlineSeconds
-	}
-}
-
-func (r *DeploymentConfigReconciler) ensurePodContainers(updated, desired []v1.Container) {
-	updatedContainerMap := map[string]*v1.Container{}
-	for idx := range updated {
-		container := &updated[idx]
-		updatedContainerMap[container.Name] = container
-	}
-
-	for idx := range desired {
-		desiredContainer := &desired[idx]
-		if updatedContainer, ok := updatedContainerMap[desiredContainer.Name]; ok {
-			desiredContainer.Image = updatedContainer.Image
+func DeploymentConfigReconcileReplicas(desired, existing *appsv1.DeploymentConfig, logger logr.Logger) bool {
```

Comment:
About the reconciliation of replicas, we have to take into account that currently:
· When HA option in the APIManager CR is NOT set, the desired replicas are set to 1
· When HA option in the APIManager CR is set, the desired replicas are set to 2, with the exception of DBs that cannot have multiple instances at the same time due to they are not compatible, where we keep it to 1.

If we reconcile replicas with the exact specified number in "desired", the user will not be able to change the number of replicas as of now. I think the solution is not good enough at this moment and we should discuss what are we going to do with this aspect.

Reply:
HAck implemented to enable manually scaling of replicas.